_posts/2020-09-01-weather-prediction/weather_prediction_deep_learning.Rmd

```diff
@@ -0,0 +1,695 @@
+
+```{r}
+deg2rad <- function(d) {
+  (d / 180) * pi
+}
+
+lats <- tidync("geopotential_500/geopotential_500hPa_2015_5.625deg.nc")$transforms$lat %>%
+  select(lat) %>%
+  pull()
+
+lat_weights <- cos(deg2rad(lats))
+lat_weights <- lat_weights / mean(lat_weights)
+
+weighted_rmse <- function(forecast, ground_truth) {
+  error <- (forecast - ground_truth) ^ 2
+  for (i in seq_along(lat_weights)) {
+    error[, i, ,] <- error[, i, ,] * lat_weights[i]
+  }
+  apply(error, 4, mean) %>% sqrt()
+}
+```
+
+#### Baseline 1: Weekly climatology
+
+In general, climatology refers to long-term averages, computed over defined time ranges. Here, we first calculate weekly
+averages based on the training set. These averages are then used to forecast the variables in question for the time period
+used as test set.
+
+Step one makes use of `tidync`, `ncmeta`, `RNetCDF` and `lubridate` to compute weekly averages for 2015, following the [ISO
+week date system](https://en.wikipedia.org/wiki/ISO_week_date).
+
+```{r}
+train_file <- "geopotential_500/geopotential_500hPa_2015_5.625deg.nc"
+
+times_train <- (tidync(train_file) %>% activate("D2") %>% hyper_array())$time
+
+time_unit_train <- ncmeta::nc_atts(train_file, "time") %>%
+  tidyr::unnest(cols = c(value)) %>%
+  dplyr::filter(name == "units")
+
+time_parts_train <- RNetCDF::utcal.nc(time_unit_train$value, times_train)
+
+iso_train <- ISOdate(
+  time_parts_train[, "year"],
+  time_parts_train[, "month"],
+  time_parts_train[, "day"],
+  time_parts_train[, "hour"],
+  time_parts_train[, "minute"],
+  time_parts_train[, "second"]
+)
+
+isoweeks_train <- map(iso_train, isoweek) %>% unlist()
+
+train_by_week <- apply(train_all, c(2, 3, 4), function(x) {
+  tapply(x, isoweeks_train, function(y) {
+    mean(y)
+  })
+})
+
+dim(train_by_week)
+```
+
+    53 32 64 2
+
+Step two then runs through the test set, mapping dates to corresponding ISO weeks and associating the weekly averages from the
+training set:
+
+```{r}
+test_file <- "geopotential_500/geopotential_500hPa_2017_5.625deg.nc"
+
+times_test <- (tidync(test_file) %>% activate("D2") %>% hyper_array())$time
+
+time_unit_test <- ncmeta::nc_atts(test_file, "time") %>%
+  tidyr::unnest(cols = c(value)) %>%
+  dplyr::filter(name == "units")
+
+time_parts_test <- RNetCDF::utcal.nc(time_unit_test$value, times_test)
+
+iso_test <- ISOdate(
+  time_parts_test[, "year"],
+  time_parts_test[, "month"],
+  time_parts_test[, "day"],
+  time_parts_test[, "hour"],
+  time_parts_test[, "minute"],
+  time_parts_test[, "second"]
+)
+
+isoweeks_test <- map(iso_test, isoweek) %>% unlist()
+
+climatology_forecast <- test_all
+
+for (i in 1:dim(climatology_forecast)[1]) {
+  week <- isoweeks_test[i]
+  lookup <- train_by_week[week, , , ]
+  climatology_forecast[i, , ,] <- lookup
+}
+
+```
+
+For this baseline, the latitudinally-weighted RMSE amounts to roughly 98 for geopotential and 4 for temperature.
```

Comment:
I think there's a typo in 98? Shouldn't it be 980?

Reply:
thanks, right! i had changed geopotential height to geopotential, now i need to add the last digit back in...